### PR TITLE
fix(webhook): evaluate SSH URLs from webhook payloads

### PR DESCRIPTION
**Pull request by d3flex**
- Use `parsePayload` to include giturl and allow the outer loop to process git along with https url
- Apply fix to all the providers for consistency
- dedup payload URLs and skip empty SSH paths
- update and refactor webhook_test.go to cover each edge case from `sshURLToParsable` and prevent regression

Refers to #3720

### CODE REVIEW

**Comment by kkaempf** [comment]
This is missing tests 🤔

**Comment by d3flex** [comment]
> This is missing tests 🤔

I thought that covering some input with the `sshURLToParsable` and add one case which has git repo in `TestGitRepoURLMatch` will do. what test do you have in mind?

**Comment by d3flex** [comment]
> This is missing tests 🤔

I rafactor the pkg/webhook/webhook_test.go to rectify the missing test cases. Those are now should cover:

- github.PushPayload
- gitlab.PushEventPayload
- gitlab.TagEventPayload
- gogsclient.PushPayload

File: pkg/webhook/webhook.go [hunk 1]
```diff
@@ -99,12 +99,23 @@ func (w *Webhook) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
...
 			w.logAndReturn(rw, err)
 			return
 		}
+		key := u.Hostname() + "\x00" + u.EscapedPath()
+		if _, ok := seen[key]; ok {
+			continue
+		}
+		seen[key] = struct{}{}
+
 		path := strings.Replace(regexp.QuoteMeta(u.EscapedPath()[1:]), `/_git/`, `(/_git)?/`, 1)
```
Comment (p-se) [inline pkg/webhook/webhook.go:117]: If there is no path provided, `u.EscapedPath()[1:]` will panic. Maybe we can fix this is with this PR as well?

File: pkg/webhook/webhook.go [hunk 2]
```diff
@@ -99,12 +99,23 @@ func (w *Webhook) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// A payload can yield several URLs (e.g. a web URL and an SSH URL) that
+	// resolve to the same host and path, most commonly when the SSH host equals
+	// the web host (github.com). Track those with `seen` so a matching GitRepo is
+	// processed once per request instead of letting it be processed repeatedly.
+	seen := make(map[string]struct{}, len(repoURLs))
```
Comment (p-se) [inline pkg/webhook/webhook.go]: Tracking and deduplicating by GitRepo resource we've already updated could be cleaner.

**Review by p-se — COMMENTED** [review]
It looks like for scp-form URLs the `.git` suffix is stripped which makes it optional, but for scheme-form payloads (gitlab on a non-22 SSH port, e.g. `ssh://git@gitlab.com:2222/owner/repo.git`) they are required and don't match spec.repo URLs without it (but scp-form URLs would).

File: pkg/webhook/webhook.go [hunk 3]
```diff
@@ -281,14 +321,23 @@ func parsePayload(payload any) (revision, branch, tag string, repoURLs []string)
 		branch, tag = getBranchTagFromRef(t.Ref)
 		revision = t.After
 		repoURLs = append(repoURLs, t.Repository.HTMLURL)
+		if sshURL := sshURLToParsable(t.Repository.SSHURL); sshURL != "" {
+			repoURLs = append(repoURLs, sshURL)
+		}
```
[the remark this answers is not in the document]
Comment (d3flex) [reply pkg/webhook/webhook.go:331]: no change

File: pkg/webhook/webhook.go (hunk 1 above)
Comment (d3flex) [reply pkg/webhook/webhook.go:117]: done - a small check preventing a panic letting the loop continue applied

File: pkg/webhook/webhook.go (hunk 2 above)
Comment (d3flex) [reply pkg/webhook/webhook.go]: after our call I removed my approach and I adapted to make use of the `gitrepo.NamespacedName` type. definetely an improvement. thanks for review and assist

File: pkg/webhook/webhook.go [hunk 4]
```diff
@@ -272,6 +286,35 @@ func getBranchTagFromRef(ref string) (string, string) {
...
 }
 
+// sshURLToParsable converts an scp-style git remote such as
+// "git@github-ssh.example.com:owner/repo.git" into an "ssh://" URL that
+// net/url can parse, so the webhook matching loop can extract its hostname and
+// path.
+func sshURLToParsable(raw string) string {
+	if raw == "" {
+		return ""
+	}
```
Comment (0xavi0) [inline pkg/webhook/webhook.go:296]:
```suggestion
    }
    // Drop a trailing ".git" so the "(\.git)?$" in the match regexp stays
    // optional and matches GitRepos configured with or without the suffix.
    // This must happen before the "://" check so it applies to both scp-style
    // remotes and ready-made "ssh://" URLs.
    raw = strings.TrimSuffix(raw, ".git")
```

File: pkg/webhook/webhook.go [hunk 5]
```diff
@@ -272,6 +286,35 @@ func getBranchTagFromRef(ref string) (string, string) {
...
+	if !ok {
+		return ""
+	}
+	host, path, ok := strings.Cut(rest, ":")
+	if !ok {
+		return ""
+	}
+	// Drop a trailing ".git" so the "(\.git)?$" in the match regexp stays
+	// optional and matches GitRepos configured with or without the suffix.
+	path = strings.TrimSuffix(path, ".git")
```
Comment (0xavi0) [inline pkg/webhook/webhook.go]:
```suggestion
```

I believe trimming the suffix before makes the following cases work fine:

### Payload
git_ssh_url = ssh://git@gitlab.example.com:2222/group/repo.git

### GitReposURL
ssh://git@gitlab.example.com:2222/group/repo
git@gitlab.example.com:group/repo

File: pkg/webhook/webhook.go (hunk 4 above)
Comment (0xavi0) [reply pkg/webhook/webhook.go:296]: See next comment

File: pkg/webhook/webhook.go (hunk 5 above)
Comment (d3flex) [reply pkg/webhook/webhook.go]: done

**Review by 0xavi0 — APPROVED** [review]

**Review by p-se — APPROVED** [review]
